src/sinks/databend/api.rs

```diff
@@ -0,0 +1,221 @@
+use std::collections::BTreeMap;
+
+use bytes::Bytes;
+use http::Request;
+use http::StatusCode;
+use hyper::Body;
+use serde::{Deserialize, Serialize};
+
+use crate::{http::Auth, http::HttpClient, sinks::util::UriSerde};
+
+use super::error::DatabendError;
+
+#[derive(Serialize, Debug)]
+pub(super) struct StageAttachment {
+    location: String,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    file_format_options: Option<BTreeMap<String, String>>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    copy_options: Option<BTreeMap<String, String>>,
+}
+
+#[derive(Serialize, Debug)]
+pub(super) struct DatabendHttpRequest {
+    sql: String,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    stage_attachment: Option<StageAttachment>,
+}
+
+impl DatabendHttpRequest {
+    pub const fn new(sql: String) -> Self {
+        Self {
+            sql,
+            stage_attachment: None,
+        }
+    }
+
+    pub fn add_stage_attachment(
+        &mut self,
+        location: String,
+        file_format_options: Option<BTreeMap<String, String>>,
+        copy_options: Option<BTreeMap<String, String>>,
+    ) {
+        self.stage_attachment = Some(StageAttachment {
+            location,
+            file_format_options,
+            copy_options,
+        });
+    }
+}
+
+#[cfg(all(test, feature = "databend-integration-tests"))]
+#[derive(Deserialize, Debug)]
+pub(super) struct DatabendHttpResponseSchemaField {
+    pub name: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub(super) struct DatabendHttpResponseError {
+    pub code: u16,
+    pub message: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub(super) struct DatabendHttpResponse {
+    #[cfg(all(test, feature = "databend-integration-tests"))]
+    pub schema: Vec<DatabendHttpResponseSchemaField>,
+    pub data: Vec<Vec<String>>,
+    pub error: Option<DatabendHttpResponseError>,
+    pub next_uri: Option<String>,
+}
+
+#[derive(Debug)]
+pub(super) struct DatabendPresignedResponse {
+    pub method: String,
+    pub headers: BTreeMap<String, String>,
+    pub url: String,
+}
+
+#[derive(Clone)]
+pub(super) struct DatabendAPIClient {
+    client: HttpClient,
+    endpoint: UriSerde,
+    auth: Option<Auth>,
+}
+
+impl DatabendAPIClient {
+    pub(super) const fn new(client: HttpClient, endpoint: UriSerde, auth: Option<Auth>) -> Self {
+        Self {
+            client,
+            endpoint,
+            auth,
+        }
+    }
+
+    pub(super) fn get_protocol(&self) -> &str {
+        self.endpoint.uri.scheme_str().unwrap_or("http")
+    }
+
+    pub(super) fn get_endpoint(&self) -> &str {
+        self.endpoint.uri.host().unwrap_or("unknown")
+    }
+
+    fn get_page_url(&self, next_uri: &str) -> Result<String, DatabendError> {
+        let api_uri = self.endpoint.append_path(next_uri)?;
+        Ok(api_uri.to_string())
+    }
+
+    fn get_query_url(&self) -> Result<String, DatabendError> {
```

Comment:
Re terminology: wouldn't this method be referring to the endpoint and the `get_endpoint` method above be referring to the host?